config/reader.go

```diff
@@ -49,31 +60,130 @@ func getAgentInfo() (*apppb.AgentInfo, error) {
 	}, nil
 }
 
-var viamDotDir = filepath.Join(os.Getenv("HOME"), ".viam")
+func updatePlaceholdersInFile(bytes []byte, packageMap map[string]string) []byte {
+	// updates all the filepath placeholders with what we think they should be based on the packages
+	updatedBytes := placeholderRegexp.ReplaceAllFunc(bytes, func(b []byte) []byte {
+		updatedPath, ok := packageMap[verifyPlaceholder(string(b))]
+		if !ok {
+			return b
+		}
 
-func getCloudCacheFilePath(id string) string {
-	return filepath.Join(viamDotDir, fmt.Sprintf("cached_cloud_config_%s.json", id))
+		return []byte(updatedPath)
+	})
+
+	return updatedBytes
+}
+
+// UpdatePlaceholdersInConfig is used to replace placeholders in a config struct.
+func UpdatePlaceholdersInConfig(config *Config) error {
```

Comment:
This feels... fragile. Converting an entire config to bytes to do token replacement, then reversing that... What if this is ever used on a live config containing structs with pointers and such? E.g. ProcessConfig things that might have a logwriter, or a cloud config with a net.Listener? As we have many structs with json set to '-' (field will be skipped on Marshall to json) this can completely remove data from a working config. Is it possible to make this a private method at least?

Reply:
The reasoning for requiring this config is for the gRPC that pulls a config directly from the cloud with placeholders in it. In theory, this should never be used on a config that contains structs with pointers etc.. Potentially would be a better idea to tie it more closely to that gRPC request so that its clear that this should not be used later on in live configs. 

This reasoning may also be why the Walker code was used in the first place. What do you think of that?

Reply:
I changed this to be called directly in the `getFromCloudGRPC` call so its explicitly tied there. We definitely do need to replace these placeholders, do you think that this would work? Not sure if there is any other way except if the robotservice returned a string version of the config rather than a struct.